src/transaction/transaction_group_complete_manager.cpp

```diff
@@ -39,55 +46,167 @@ namespace cubtx
 
     on_register_transaction ();
 
+    er_log_debug (ARG_FILE_LINE, "group_complete_manager::register_transaction: (tran_index=%d, MVCCID=%llu, " \
+		  "tran_state=%d, current group size = %d)\n", tran_index, (long long int) mvccid, (int) state,
+		  m_current_group.get_container ().size ());
+
     return m_current_group_id;
   }
 
   //
-  // wait_for_complete_mvcc waits for MVCC complete event on specified group_id.
+  // complete_mvcc complete transactions MVCC.
   //
-  void group_complete_manager::wait_for_complete_mvcc (id_type group_id)
+  void group_complete_manager::complete_mvcc (id_type group_id)
   {
     if (group_id < m_latest_closed_group_id)
       {
 	/* Already advanced to next group. No need to acquire mutex. */
+	er_log_debug (ARG_FILE_LINE, "group_complete_manager::complete_mvcc: (tran_index = %d, group_id = %llu)\n",
+		      LOG_FIND_THREAD_TRAN_INDEX (&cubthread::get_entry ()), group_id);
 	return;
       }
 
-    std::unique_lock<std::mutex> ulock (m_group_complete_mutex);
-    /* TODO - consider stop and optimize next call */
-    m_group_complete_condvar.wait (ulock, [&] {return is_group_mvcc_completed (group_id);});
+#if defined (SERVER_MODE)
+    if (BO_IS_SERVER_RESTARTED ())
+      {
+	if (group_id == m_latest_closed_group_id)
+	  {
+	    if (is_latest_closed_group_mvcc_completed ())
+	      {
+		/* Completed mvcc for group_id. No need to acquire mutex. */
+		assert (group_id <= m_latest_closed_group_id);
+		er_log_debug (ARG_FILE_LINE, "group_complete_manager::complete_mvcc: (tran_index = %d, group_id = %llu)\n",
+			      LOG_FIND_THREAD_TRAN_INDEX (&cubthread::get_entry ()), group_id);
+		return;
+	      }
+	  }
+
+	/* Waits for MVCC complete event on specified group_id, even in async case. */
+	std::unique_lock<std::mutex> ulock (m_group_complete_mutex);
+	/* TODO - consider stop and optimize next call */
+	m_group_complete_condvar.wait (ulock, [&] {return is_group_mvcc_completed (group_id);});
+
+	er_log_debug (ARG_FILE_LINE, "group_complete_manager::complete_mvcc: (tran_index = %d, group_id = %llu)\n",
```

Comment:
I think is better to unlock the mutex while er_log_debug

Reply:
Is better, but we don't want to change the order of messages (a transaction added to next group and then the current group closed, for instance).